user/tests/wiring/no_fixture/acm.cpp

```diff
@@ -0,0 +1,155 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
+ */
+#include "application.h"
+#include "unit-test/unit-test.h"
+
+const uint32_t WIFI_CONNECT_TIMEOUT_MS = 5 * 60 * 1000;
+const uint32_t DISCONNECT_TIMEOUT_MS = 5 * 60 * 1000;
+
+test(ACM_01_PREPARE) {
+    Particle.disconnect();
+
+#if HAL_PLATFORM_WIFI
+    WiFi.disconnect();
+#endif // HAL_PLATFORM_WIFI
+#if HAL_PLATFORM_CELLULAR
+    Cellular.disconnect();
+#endif // HAL_PLATFORM_CELLULAR
+#if HAL_PLATFORM_ETHERNET
+    Ethernet.disconnect();
+#endif // HAL_PLATFORM_ETHERNET
+
+    assertEqual(Network, Particle.connectionInterface());
+}
+
+test(ACM_02_particle_connect_uses_preferred_network) {
+    auto expectedNetwork = Network;
+
+#if HAL_PLATFORM_WIFI && !HAL_PLATFORM_WIFI_SCAN_ONLY
+    expectedNetwork = WiFi;
+    WiFi.prefer();
+    WiFi.connect();
+    assertTrue(waitFor(WiFi.ready, WIFI_CONNECT_TIMEOUT_MS));
+#endif // HAL_PLATFORM_WIFI && !HAL_PLATFORM_WIFI_SCAN_ONLY
+
+#if HAL_PLATFORM_CELLULAR
+    expectedNetwork = Cellular;
+    Cellular.prefer();
+    Cellular.connect();
+    assertTrue(waitFor(Cellular.ready, HAL_PLATFORM_CELLULAR_CONN_TIMEOUT));
+#endif // HAL_PLATFORM_CELLULAR
+
+    Particle.connect();
+    assertTrue(waitFor(Particle.connected, HAL_PLATFORM_MAX_CLOUD_CONNECT_TIME));
+    assertEqual(Particle.connectionInterface(), expectedNetwork);
+}
+
+
+#if PLATFORM_ID == PLATFORM_MSOM
+test(ACM_03_prefer_one_network_only) {
+	WiFi.prefer();
+	assertFalse(Cellular.isPreferred());
+	Cellular.prefer();
+	assertFalse(WiFi.isPreferred());
+}
+
+test(ACM_04_prefer_moves_cloud_connection_when_set) {
+    // Default preference
+	Network.prefer();
+	
+    // Connect all interfaces
+    Network.connect();
+    Particle.connect();
+    assertTrue(waitFor(Particle.connected, HAL_PLATFORM_MAX_CLOUD_CONNECT_TIME));
+
+    auto connectedNetwork = Particle.connectionInterface();
+
+    // Prefer whatever network we did not connect on in order to force the cloud connection to switch
+    if (connectedNetwork == Cellular) {
+	    assertTrue(waitFor(WiFi.ready, WIFI_CONNECT_TIMEOUT_MS));
+    	WiFi.prefer();
+    } else {
+    	assertTrue(waitFor(Cellular.ready, HAL_PLATFORM_CELLULAR_CONN_TIMEOUT));
+    	Cellular.prefer();
+    }
+
+    // Verify that we disconnect and reconnect on the newly preferred network
+    assertTrue(waitForNot(Particle.connected, DISCONNECT_TIMEOUT_MS));
```

Comment:
I guess it's not likely that we miss the disconnect event, but I'm thinking it might be better to monitor system networking events to ensure we catch the disconnect and reconnect events.